lib/hotel.rb

```diff
@@ -0,0 +1,168 @@
+    return search_reservations
+  end
+
+  def self.available_rooms(begin_date, end_date)
+    begin_search = Date.new(begin_date[0].to_i, begin_date[1].to_i, begin_date[2].to_i)
+    end_search = Date.new(end_date[0].to_i, end_date[1].to_i, end_date[2].to_i)
+    unavailable_rooms = []
+    all_reservations = self.all_reservations
+    all_reservations.each do |reservation|
+      if (begin_search >= reservation.check_in) && (begin_search < reservation.check_out) && (end_search >= reservation.check_in) && (end_search <= reservation.check_out)
+        unavailable_rooms<< reservation.room
+      end
+    end
+    available_rooms = HOTEL_ROOMS - unavailable_rooms
+    return available_rooms
+  end
+
+  def self.blocked_rooms(begin_date, end_date)
+    begin_search = Date.new(begin_date[0].to_i, begin_date[1].to_i, begin_date[2].to_i)
+    end_search = Date.new(end_date[0].to_i, end_date[1].to_i, end_date[2].to_i)
+    blocked_rooms = []
+    all_the_blocks = self.all_blocks
+    all_the_blocks.each do |block|
+      if (begin_search >= block.check_in) && (begin_search < block.check_out) && (end_search >= block.check_in) && (end_search <= block.check_out)
+        block.rooms.each do |room|
+          blocked_rooms << room
+        end
+      end
+    end
+    return blocked_rooms
+  end
+
+  def self.truly_available(begin_date, end_date)
+    available_rooms = self.available_rooms(begin_date, end_date)
+    blocked_rooms = self.blocked_rooms(begin_date, end_date)
+    truly_available = available_rooms - blocked_rooms
+    return truly_available
+  end
+
+
+  def self.reserve_room(begin_date, end_date)
+    # begin_reservation = Date.new(begin_date[0], begin_date[1], begin_date[2])
+    # end_reservation = Date.new(end_date[0], end_date[1], end_date[2])
+    available_rooms = self.available_rooms(begin_date, end_date)
+    raise ArgumentError.new "No rooms are available for this date range" if available_rooms.length == 0
+    return Hotel::Reservation.new(10, available_rooms[0].id, begin_date, end_date)
+  end
+
+  def self.reserve_block(rate, num_of_rooms, check_in_date, check_out_date, input_block_id)
+    available_rooms = truly_available(check_in_date, check_out_date)
+    raise ArgumentError.new "Not enough rooms are available for this date range" if available_rooms.length < num_of_rooms.to_i
+    return Hotel::Block.new(rate, num_of_rooms, check_in_date, check_out_date, input_block_id)
+  end
+
+  def self.block_available(block_id)
+    block = self.find_block(block_id)
+    rooms = block.rooms.length
+    taken = self.find_reservation_by_block_id(block_id)
+    return rooms - taken.length
+  end
+
+  def self.reserve_block_room(block_id)
+    raise ArgumentError.new "Invalid input. Please enter Block ID as an Integer" if !(block_id.is_a? Integer)
+    block = self.find_block(block_id)
+    raise ArgumentError.new "Block ID is not found" if !(block.is_a? Hotel::Block)
+    # block.add_rooms
+    availability = block_available(block_id)
```

Comment:
in this method, you've already found the block object and stored it in `block`, but then you redo that work in calling `block_available` and passing in the block_id instead of the block itself.

I would refactor this so that it relies on block more, and lets block return internal information about itself